File: jar/src/main/java/jline/api/mc/Ctmc_stmonotone.java
package jline.api.mc;

import jline.util.Pair;
import jline.util.matrix.Matrix;

public final class Ctmc_stmonotone {
    private Ctmc_stmonotone() {}

    /**
     * Computes the stochastically monotone upper bound for a CTMC.
     */
    public static Matrix ctmc_stmonotone(Matrix Q) {
        double maxAbsVal = 0.0;
        for (int i = 0; i < Q.getNumRows(); i++) {
            for (int j = 0; j < Q.getNumCols(); j++) {
                maxAbsVal = Math.max(maxAbsVal, Math.abs(Q.get(i, j)));
            }
        }

        Pair<Matrix, Double> rand = Ctmc_randomization.ctmc_randomization(Q, maxAbsVal);
        Matrix P = rand.getLeft();

        Matrix P_stochastic = Dtmc_makestochastic.dtmc_makestochastic(P);

        Matrix Pub = dtmc_stmonotone(P_stochastic);

        Matrix Qub = Ctmc_makeinfgen.ctmc_makeinfgen(Pub);

        return Qub;
    }

    /**
     * Implementation of the dtmc_stmonotone algorithm.
     */
    public static Matrix dtmc_stmonotone(Matrix P) {
        int n = P.length();
        Matrix Q = new Matrix(n, n);

        Q.set(0, n - 1, P.get(0, n - 1));

        for (int i = 1; i < n; i++) {
            Q.set(i, n - 1, Math.max(Q.get(i - 1, n - 1), P.get(i, n - 1)));
        }

        for (int l = n - 2; l >= 0; l--) {
            Q.set(0, l, P.get(0, l));

            for (int i = 1; i < n; i++) {
                double sumQ_prev = 0.0;
                for (int k = l; k < n; k++) {
                    sumQ_prev += Q.get(i - 1, k);
                }

                double sumP = 0.0;
                for (int k = l; k < n; k++) {
                    sumP += P.get(i, k);
                }

                double sumQ_right = 0.0;
                for (int k = l + 1; k < n; k++) {
                    sumQ_right += Q.get(i, k);
                }

                Q.set(i, l, Math.max(sumQ_prev, sumP) - sumQ_right);
            }
        }

        return Q;
    }
}

File: jar/src/main/kotlin/jline/api/mc/Ctmc_stmonotone.kt
package jline.api.mc

import jline.util.matrix.Matrix
import kotlin.math.abs
import kotlin.math.max

/**
 * Computes the stochastically monotone upper bound for a CTMC
 * Based on Forneau, Pekergin - An Algorithmic Approach to Stochastic Bounds, Performance 2002
 *
 * @param Q infinitesimal generator matrix
 * @return Qub stochastically monotone upper bound infinitesimal generator
 */
fun ctmc_stmonotone(Q: Matrix): Matrix {
    // Find the maximum absolute value in Q for randomization
    var maxAbsVal = 0.0
    for (i in 0 until Q.getNumRows()) {
        for (j in 0 until Q.getNumCols()) {
            maxAbsVal = max(maxAbsVal, abs(Q.get(i, j)))
        }
    }
    
    // Apply randomization to convert CTMC to DTMC
    val (P, _) = ctmc_randomization(Q, maxAbsVal)
    
    // Make sure P is stochastic
    val P_stochastic = dtmc_makestochastic(P)
    
    // Apply the stochastically monotone algorithm to the DTMC
    val Pub = dtmc_stmonotone(P_stochastic)
    
    // Convert back to infinitesimal generator
    val Qub = ctmc_makeinfgen(Pub)
    
    return Qub
}

/**
 * Implementation of the dtmc_stmonotone algorithm
 * Forneau, Pekergin - An Algorithmic Approach to Stochastic Bounds, Performance 2002 
 * Algorithm 1
 *
 * @param P transition probability matrix
 * @return Q stochastically monotone upper bound matrix
 */
fun dtmc_stmonotone(P: Matrix): Matrix {
    val n = P.length()
    val Q = Matrix(n, n)
    
    // Initialize last column
    Q.set(0, n - 1, P.get(0, n - 1))
    
    for (i in 1 until n) {
        Q.set(i, n - 1, max(Q.get(i - 1, n - 1), P.get(i, n - 1)))
    }
    
    // Fill remaining columns from right to left
    for (l in n - 2 downTo 0) {
        Q.set(0, l, P.get(0, l))
        
        for (i in 1 until n) {
            // Calculate sum(Q(i-1, l:n))
            var sumQ_prev = 0.0
            for (k in l until n) {
                sumQ_prev += Q.get(i - 1, k)
            }
            
            // Calculate sum(P(i, l:n))
            var sumP = 0.0
            for (k in l until n) {
                sumP += P.get(i, k)
            }
            
            // Calculate sum(Q(i, (l+1):n))
            var sumQ_right = 0.0
            for (k in l + 1 until n) {
                sumQ_right += Q.get(i, k)
            }
            
            Q.set(i, l, max(sumQ_prev, sumP) - sumQ_right)
        }
    }
    
    return Q
}
/**
 * CTMC stmonotone algorithms
 */
@Suppress("unused")
class CtmcStmonotoneAlgo {
    companion object {
        // Class documentation marker for Dokka
    }
}